src/shared/pager.c

```diff
@@ -165,25 +165,34 @@ int pager_open(PagerFlags flags) {
                 }
 
                 /* People might invoke us from sudo, don't needlessly allow less to be a way to shell out
-                 * privileged stuff. */
-                r = getenv_bool("SYSTEMD_LESSSECURE");
-                if (r == 0) { /* Remove env var if off */
-                        if (unsetenv("LESSSECURE") < 0) {
-                                log_error_errno(errno, "Failed to uset environment variable LESSSECURE: %m");
-                                _exit(EXIT_FAILURE);
-                        }
-                } else {
-                        /* Set env var otherwise */
-                        if (r < 0)
-                                log_warning_errno(r, "Unable to parse $SYSTEMD_LESSSECURE, ignoring: %m");
+                 * privileged stuff. If the user set $SYSTEMD_PAGERSECURE, trust their configuration of the
+                 * pager. If they didn't, use secure mode when under sudo. If $SYSTEMD_PAGERSECURE wasn't
+                 * explicitly set, only use the pager we know to be good. */
+                int secure = getenv_bool("SYSTEMD_PAGERSECURE");
+                bool trust_pager = secure >= 0;
+                if (secure == -ENXIO)
+                        secure = !!getenv("SUDO_COMMAND");
```

Comment:
I assume you mean `sd_pid_get_owner_uid() == geteuid()`. I like the idea, but it seems broken by the move to starting things under user@.service: sd_pid_get_owner_uid() returns ENODATA for the majority of user processes started from a graphical session. As a fallback it would be possible to use sd_pid_get_unit() and query PID1 for User= setting of the returned unit. But frankly, this makes a rather long chain of steps necessary to make a security decision. Those steps are non-atomic, but more importantly, any of them can fail if we the machine is not fully up or some services are not running. And if those steps fail, then we need to assume some default and things become messy. Communicating with pid1 also has the disadvantage that is brings it into operations where it wasn't required before. E.g. `systemctl --root=/ some-op` stops being independent of pid1 now, which could be a problem if the user is debugging a pid1 problem.

I think the original approach of checking `$SUDO_COMMAND` is more practical — if the variable is set, we are quite certain that we're under sudo. And it is a strict improvement on status quo without any guesswork. So we improve on the common case but otherwise things stay as they were.

I'll push another commit to do the same for pkexec. It sets `$PKEXEC_UID` which we can use for the same purpose.

Reply:
sd_pid_get_owner_uid() should work for everything under the user's slice, and that includes stuff run from `user@.service`. If that doesn#t work, then that'd be a bug

Reply:
> I think the original approach of checking `$SUDO_COMMAND` is more practical — if the variable is set, we are quite certain that we're under sudo. And it is a strict improvement on status quo without any guesswork. So we improve on the common case but otherwise things stay as they were.

Humm, that's not really the case. env vars propagate down the tree. It gets eaten up at some places where we'd not recognize things. And it gets passed where it shouldn't. I think the sd_pid_get_owner_uid() check is much more correct: it doesn't actually check *how* you got where you got. It just checks that what we are now matches what you started as.

Reply:
I tried to apply the discussed patch on a remote machine, so I connected through ssh, and all sub-processes were running inside "system.slice", not "user-xxxx.slice", as a result, there is not possible to determine pid with the function `sd_pid_get_owner_uid()`. 

Why didn't you use the control of either `SUDO_COMMAND` or `SUDO_UID` environment variables ?